### PR TITLE
docs(builtin): fix and annotate language blocks

### PR DESCRIPTION
**Pull request by lewis6991**

### CODE REVIEW

File: runtime/doc/builtin.txt [hunk 1]
```diff
@@ -612,17 +633,18 @@ byteidx({expr}, {nr} [, {utf16}])				*byteidx()*
 		in bytes is returned.
 		See |charidx()| and |utf16idx()| for getting the character and
 		UTF-16 index respectively from the byte index.
-		Examples: >
+		Examples: >vim
 			echo byteidx('a😊😊', 2)	returns 5
 			echo byteidx('a😊😊', 2, 1)	returns 1
 			echo byteidx('a😊😊', 3, 1)	returns 5
```
Comment (clason) [inline runtime/doc/builtin.txt:639]:
```suggestion
			echo byteidx('a😊😊', 2)	    " returns 5
			echo byteidx('a😊😊', 2, 1)	" returns 1
			echo byteidx('a😊😊', 3, 1)	" returns 5
```

File: runtime/doc/builtin.txt [hunk 2]
```diff
@@ -361,39 +371,41 @@ atan({expr})							*atan()*
...
+		Can also be used as a |method|: >vim
 			Compute()->atan2(1)
+<
 
 blob2list({blob})					*blob2list()*
 		Return a List containing the number value of each byte in Blob
-		{blob}.  Examples: >
+		{blob}.  Examples: >vim
 			blob2list(0z0102.0304)	returns [1, 2, 3, 4]
 			blob2list(0z)		returns []
```
Comment (clason) [inline runtime/doc/builtin.txt:404]:
```suggestion
			blob2list(0z0102.0304)	" returns [1, 2, 3, 4]
			blob2list(0z)           " returns []
```

File: runtime/doc/builtin.txt [hunk 3]
```diff
@@ -511,13 +528,13 @@ bufname([{buf}])						*bufname()*
...
-<		Can also be used as a |method|: >
+<		Can also be used as a |method|: >vim
 			echo bufnr->bufname()
 
 <		If the buffer doesn't exist, or doesn't have a name, an empty
-		string is returned. >
+		string is returned. >vim
 			bufname("#")		alternate buffer name
 			bufname(3)		name of buffer 3
 			bufname("%")		name of current buffer
```
Comment (clason) [inline runtime/doc/builtin.txt:540]:
```suggestion
			bufname("#")		" alternate buffer name
			bufname(3)		" name of buffer 3
			bufname("%")		" name of current buffer
```

File: runtime/doc/builtin.txt [hunk 4]
```diff
@@ -704,7 +727,7 @@ chansend({id}, {data})					*chansend()*
 
 char2nr({string} [, {utf8}])				*char2nr()*
 		Return Number value of the first char in {string}.
-		Examples: >
+		Examples: >vim
 			char2nr(" ")		returns 32
 			char2nr("ABC")		returns 65
 			char2nr("á")		returns 225
```
Comment (clason) [inline runtime/doc/builtin.txt:733]:
```suggestion
			char2nr(" ")		" returns 32
			char2nr("ABC")		" returns 65
			char2nr("á")		" returns 225
```

File: runtime/doc/builtin.txt [hunk 5]
```diff
@@ -736,11 +760,11 @@ charcol({expr} [, {winid}])				*charcol()*
 		position given with {expr} instead of the byte position.
 
 		Example:
-		With the cursor on '세' in line 5 with text "여보세요": >
+		With the cursor on '세' in line 5 with text "여보세요": >vim
 			charcol('.')		returns 3
 			col('.')		returns 7
```
Comment (clason) [inline runtime/doc/builtin.txt:765]:
```suggestion
			charcol('.')		" returns 3
			col('.')		" returns 7
```

File: runtime/doc/builtin.txt [hunk 6]
```diff
@@ -771,14 +795,15 @@ charidx({string}, {idx} [, {countcc} [, {utf16}]])	*charidx()*
 		from the character index and |utf16idx()| for getting the
 		UTF-16 index from the character index.
 		Refer to |string-offset-encoding| for more information.
-		Examples: >
+		Examples: >vim
 			echo charidx('áb́ć', 3)		returns 1
 			echo charidx('áb́ć', 6, 1)	returns 4
 			echo charidx('áb́ć', 16)		returns -1
 			echo charidx('a😊😊', 4, 0, 1)	returns 2
```
Comment (clason) [inline runtime/doc/builtin.txt:802]:
```suggestion
			echo charidx('áb́ć', 3)		" returns 1
			echo charidx('áb́ć', 6, 1)	" returns 4
			echo charidx('áb́ć', 16)		" returns -1
			echo charidx('a😊😊', 4, 0, 1)	" returns 2
```

File: runtime/doc/builtin.txt [hunk 7]
```diff
@@ -850,7 +876,7 @@ col({expr} [, {winid}])					*col()*
 		For the screen column position use |virtcol()|.  For the
 		character position use |charcol()|.
 		Note that only marks in the current file can be used.
-		Examples: >
+		Examples: >vim
 			col(".")		column of cursor
 			col("$")		length of cursor line plus one
 			col("'t")		column of mark t
```
Comment (clason) [inline runtime/doc/builtin.txt:882]:
```suggestion
			col(".")		" column of cursor
			col("$")		" length of cursor line plus one
			col("'t")		" column of mark t
```

File: runtime/doc/builtin.txt [hunk 8]
```diff
@@ -1726,12 +1769,12 @@ expandcmd({string} [, {options}])			*expandcmd()*
 		Returns the expanded string.  If an error is encountered
 		during expansion, the unmodified {string} is returned.
 
-		Example: >
+		Example: >vim
 			:echo expandcmd('make %<.o')
 			make /path/runtime/doc/builtin.o
```
Comment (clason) [inline runtime/doc/builtin.txt:1774]:
```suggestion
<			make /path/runtime/doc/builtin.o  >vim
```

File: runtime/doc/builtin.txt [hunk 9]
```diff
@@ -1835,23 +1879,25 @@ filereadable({file})					*filereadable()*
 		expression, which is used as a String.
 		If you don't care about the file being readable you can use
 		|glob()|.
-		{file} is used as-is, you may want to expand wildcards first: >
+		{file} is used as-is, you may want to expand wildcards first: >vim
 			echo filereadable('~/.vimrc')
 			0
```
Comment (clason) [inline runtime/doc/builtin.txt:1884]:
```suggestion
<			0  >vim
```

File: runtime/doc/builtin.txt [hunk 10]
```diff
@@ -1835,23 +1879,25 @@ filereadable({file})					*filereadable()*
...
 			echo filereadable(expand('~/.vimrc'))
 			1
```
Comment (clason) [inline runtime/doc/builtin.txt:1886]:
```suggestion
<			1
```

File: runtime/doc/builtin.txt [hunk 11]
```diff
@@ -2534,12 +2593,13 @@ getcharpos({expr})					*getcharpos()*
 		of the last character.
 
 		Example:
-		With the cursor on '세' in line 5 with text "여보세요": >
+		With the cursor on '세' in line 5 with text "여보세요": >vim
 			getcharpos('.')		returns [0, 5, 3, 0]
 			getpos('.')		returns [0, 5, 7, 0]
```
Comment (clason) [inline runtime/doc/builtin.txt:2598]:
```suggestion
			getcharpos('.')		" returns [0, 5, 3, 0]
			getpos('.')		" returns [0, 5, 7, 0]
```

File: runtime/doc/builtin.txt [hunk 12]
```diff
@@ -2736,12 +2796,13 @@ getcursorcharpos([{winid}])			*getcursorcharpos()*
 		List is a character index instead of a byte index.
 
 		Example:
-		With the cursor on '보' in line 3 with text "여보세요": >
+		With the cursor on '보' in line 3 with text "여보세요": >vim
 			getcursorcharpos()	returns [0, 3, 2, 0, 3]
 			getcurpos()		returns [0, 3, 4, 0, 3]
```
Comment (clason) [inline runtime/doc/builtin.txt:2801]:
```suggestion
			getcursorcharpos()	" returns [0, 3, 2, 0, 3]
			getcurpos()		" returns [0, 3, 4, 0, 3]
```

File: runtime/doc/builtin.txt [hunk 13]
```diff
@@ -4437,7 +4538,7 @@ line({expr} [, {winid}])					*line()*
 		With the optional {winid} argument the values are obtained for
 		that window instead of the current window.
 		Returns 0 for invalid values of {expr} and {winid}.
-		Examples: >
+		Examples: >vim
 			line(".")		line number of the cursor
 			line(".", winid)	idem, in window "winid"
 			line("'t")		line number of mark t
```
Comment (clason) [inline runtime/doc/builtin.txt:4544]:
```suggestion
			line(".")		" line number of the cursor
			line(".", winid)	" idem, in window "winid"
			line("'t")		" line number of mark t
```

File: runtime/doc/builtin.txt [hunk 14]
```diff
@@ -4472,40 +4575,43 @@ lispindent({lnum})					*lispindent()*
...
+		Can also be used as a |method|: >vim
 			GetLnum()->lispindent()
+<
 
 list2blob({list})					*list2blob()*
 		Return a Blob concatenating all the number values in {list}.
-		Examples: >
+		Examples: >vim
 			list2blob([1, 2, 3, 4])	returns 0z01020304
 			list2blob([])		returns 0z
```
Comment (clason) [inline runtime/doc/builtin.txt:4586]:
```suggestion
			list2blob([1, 2, 3, 4])	" returns 0z01020304
			list2blob([])		" returns 0z
```

File: runtime/doc/builtin.txt [hunk 15]
```diff
@@ -4472,40 +4575,43 @@ lispindent({lnum})					*lispindent()*
...
+		Can also be used as a |method|: >vim
 			GetList()->list2blob()
+<
 
 list2str({list} [, {utf8}])				*list2str()*
 		Convert each number in {list} to a character string can
-		concatenate them all.  Examples: >
+		concatenate them all.  Examples: >vim
 			list2str([32])		returns " "
 			list2str([65, 66, 67])	returns "ABC"
```
Comment (clason) [inline runtime/doc/builtin.txt:4600]:
```suggestion
			list2str([32])		" returns " "
			list2str([65, 66, 67])	" returns "ABC"
```

File: runtime/doc/builtin.txt [hunk 16]
```diff
@@ -4472,40 +4575,43 @@ lispindent({lnum})					*lispindent()*
...
-<		The same can be done (slowly) with: >
+<		The same can be done (slowly) with: >vim
 			join(map(list, {nr, val -> nr2char(val)}), '')
 <		|str2list()| does the opposite.
 
 		UTF-8 encoding is always used, {utf8} option has no effect,
 		and exists only for backwards-compatibility.
-		With UTF-8 composing characters work as expected: >
+		With UTF-8 composing characters work as expected: >vim
 			list2str([97, 769])	returns "á"
```
Comment (clason) [inline runtime/doc/builtin.txt:4608]:
```suggestion
			list2str([97, 769])	" returns "á"
```

File: runtime/doc/builtin.txt [hunk 17]
```diff
@@ -5415,22 +5532,23 @@ msgpackparse({data})					*msgpackparse()*
...
+		Can also be used as a |method|: >vim
 			GetLnum()->nextnonblank()
+<
 
 nr2char({expr} [, {utf8}])				*nr2char()*
 		Return a string with a single character, which has the number
-		value {expr}.  Examples: >
+		value {expr}.  Examples: >vim
 			nr2char(64)		returns "@"
 			nr2char(32)		returns " "
```
Comment (clason) [inline runtime/doc/builtin.txt:5550]:
```suggestion
			nr2char(64)		" returns "@"
			nr2char(32)		" returns " "
```

File: runtime/doc/builtin.txt [hunk 18]
```diff
@@ -5415,22 +5532,23 @@ msgpackparse({data})					*msgpackparse()*
...
-<		Example for "utf-8": >
+<		Example for "utf-8": >vim
 			nr2char(300)		returns I with bow character
```
Comment (clason) [inline runtime/doc/builtin.txt:5552]:
```suggestion
			nr2char(300)		" returns I with bow character
```

File: runtime/doc/builtin.txt [hunk 19]
```diff
@@ -8081,28 +8253,30 @@ str2float({string} [, {quoted}])			*str2float()*
...
+		Can also be used as a |method|: >vim
 			let f = text->substitute(',', '', 'g')->str2float()
+<
 
 str2list({string} [, {utf8}])				*str2list()*
 		Return a list containing the number values which represent
-		each character in String {string}.  Examples: >
+		each character in String {string}.  Examples: >vim
 			str2list(" ")		returns [32]
 			str2list("ABC")		returns [65, 66, 67]
```
Comment (clason) [inline runtime/doc/builtin.txt:8269]:
```suggestion
			str2list(" ")		" returns [32]
			str2list("ABC")		" returns [65, 66, 67]
```

File: runtime/doc/builtin.txt [hunk 20]
```diff
@@ -8081,28 +8253,30 @@ str2float({string} [, {quoted}])			*str2float()*
...
 <		|list2str()| does the opposite.
 
 		UTF-8 encoding is always used, {utf8} option has no effect,
 		and exists only for backwards-compatibility.
-		With UTF-8 composing characters are handled properly: >
+		With UTF-8 composing characters are handled properly: >vim
 			str2list("á")		returns [97, 769]
```
Comment (clason) [inline runtime/doc/builtin.txt:8275]:
```suggestion
			str2list("á")		" returns [97, 769]
```

File: runtime/doc/builtin.txt [hunk 21]
```diff
@@ -8217,16 +8393,17 @@ strftime({format} [, {time}])				*strftime()*
...
 		See also |localtime()|, |getftime()| and |strptime()|.
 		The language can be changed with the |:language| command.
-		Examples: >
+		Examples: >vim
 		  :echo strftime("%c")		   Sun Apr 27 11:49:23 1997
 		  :echo strftime("%Y %b %d %X")	   1997 Apr 27 11:53:25
 		  :echo strftime("%y%m%d %T")	   970427 11:53:55
 		  :echo strftime("%H:%M")	   11:55
 		  :echo strftime("%c", getftime("file.c"))
 						   Show mod time of file.c.
```
Comment (clason) [inline runtime/doc/builtin.txt:8402]:
```suggestion
		  :echo strftime("%c")		   " Sun Apr 27 11:49:23 1997
		  :echo strftime("%Y %b %d %X")	   " 1997 Apr 27 11:53:25
		  :echo strftime("%y%m%d %T")	   " 970427 11:53:55
		  :echo strftime("%H:%M")	   " 11:55
		  :echo strftime("%c", getftime("file.c"))
						   " Show mod time of file.c.
```

File: runtime/doc/builtin.txt [hunk 22]
```diff
@@ -8237,29 +8414,30 @@ strgetchar({str}, {index})				*strgetchar()*
...
 			:let colon2 = stridx(line, ":", colon1 + 1)
 <		The search is done case-sensitive.
 		For pattern searches use |match()|.
 		-1 is returned if the {needle} does not occur in {haystack}.
 		See also |strridx()|.
-		Examples: >
+		Examples: >vim
 		  :echo stridx("An Example", "Example")	     3
 		  :echo stridx("Starting point", "Start")    0
 		  :echo stridx("Starting point", "start")   -1
```
Comment (clason) [inline runtime/doc/builtin.txt:8435]:
```suggestion
		  :echo stridx("An Example", "Example")	     " 3
		  :echo stridx("Starting point", "Start")    " 0
		  :echo stridx("Starting point", "start")   " -1
```

File: runtime/doc/builtin.txt [hunk 23]
```diff
@@ -8404,14 +8587,14 @@ strutf16len({string} [, {countcc}])			*strutf16len()*
 		Returns zero on error.
 
 		Also see |strlen()| and |strcharlen()|.
-		Examples: >
+		Examples: >vim
 		    echo strutf16len('a')		returns 1
 		    echo strutf16len('©')		returns 1
 		    echo strutf16len('😊')		returns 2
 		    echo strutf16len('ą́')		returns 1
 		    echo strutf16len('ą́', v:true)	returns 3
```
Comment (clason) [inline runtime/doc/builtin.txt:8595]:
```suggestion
		    echo strutf16len('a')		" returns 1
		    echo strutf16len('©')		" returns 1
		    echo strutf16len('😊')		" returns 2
		    echo strutf16len('ą́')		" returns 1
		    echo strutf16len('ą́', v:true)	 " returns 3
```

File: runtime/doc/builtin.txt [hunk 24]
```diff
@@ -9173,7 +9375,7 @@ utf16idx({string}, {idx} [, {countcc} [, {charidx}]])	*utf16idx()*
 		from the UTF-16 index and |charidx()| for getting the
 		character index from the UTF-16 index.
 		Refer to |string-offset-encoding| for more information.
-		Examples: >
+		Examples: >vim
 			echo utf16idx('a😊😊', 3)	returns 2
 			echo utf16idx('a😊😊', 7)	returns 4
 			echo utf16idx('a😊😊', 1, 0, 1)	returns 2
```
Comment (clason) [inline runtime/doc/builtin.txt:9381]:
```suggestion
			echo utf16idx('a😊😊', 3)	" returns 2
			echo utf16idx('a😊😊', 7)	" returns 4
			echo utf16idx('a😊😊', 1, 0, 1)	" returns 2
```

File: runtime/doc/builtin.txt [hunk 25]
```diff
@@ -9182,7 +9384,7 @@ utf16idx({string}, {idx} [, {countcc} [, {charidx}]])	*utf16idx()*
 			echo utf16idx('aą́c', 6, 1)	returns 4
 			echo utf16idx('a😊😊', 9)	returns -1
```
Comment (clason) [inline runtime/doc/builtin.txt:9385]:
```suggestion
			echo utf16idx('aą́c', 6, 1)	" returns 4
			echo utf16idx('a😊😊', 9)	" returns -1
```

**Review by clason — COMMENTED** [review]
Sorry for commenting on the generated file (but I'd have messed up the whitespace anyway...)

Does it make sense to normalize the snippets to start with or (probably easier/better) without colon?

**Comment by clason** [comment]
And don't forget the static section "2. Matching a pattern in a String" (in `./scripts/gen_eval_files.lua`)!

**Comment by clason** [comment]
Random tangent (but don't want to forget): if we're touching `builtin.txt`, we can finally fix the invalid links `|v:_null_{blob,dict,list,string}|` that need excluding in `scripts/gen_help_html.lua`.

**Comment by lewis6991** [comment]
There should only be 2 ts errors by my estimation.

**Comment by clason** [comment]
As long as `scripts/gen_help_html.lua` doesn't complain, we're good!

**Review by clason — APPROVED** [review]

**Comment by clason** [comment]
oh boy, I should have tested this first -- `:h builtin` becomes unusable with all those injections 😞

(Looks nice, tho -- as a slideshow...)

**Comment by clason** [comment]
Regarding the parsing errors: The parser expects the closing `<` marker to be at column 0 here.

**Comment by lewis6991** [comment]
> oh boy, I should have tested this first -- `:h builtin` becomes unusable with all those injections 😞
> 
> (Looks nice, tho -- as a slideshow...)
> 
> Guess we have a strong incentive to revisit the injection performance improvements now 😬 

Oh no, I didn't notice. I hope it isn't too bad.

**Comment by clason** [comment]
It's horrible. As in, I need to `kill nvim` if I do `:h builtin` because the editor becomes unresponsive for minutes.

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/24508 happens to remove many of the codeblocks in builtin.txt, so that should help performance.

**Comment by justinmk** [comment]
Should I wait on this before merging  https://github.com/neovim/neovim/pull/24508 ?

**Comment by lewis6991** [comment]
I'm undecided. The perf issues might mean we should sit on this for a bit.

Could you test this and see what it's like for you? I didn't notice any particular slowdown myself.

**Comment by clason** [comment]
> I'm undecided. The perf issues might mean we should sit on this for a bit.
> 
> Could you test this and see what it's like for you? I didn't notice any particular slowdown myself.

The perf issues I mention require the `comment` injections from nvim-treesitter; this means you'll have to have both `vim` and `comment` parsers installed using that plugin.

**Comment by lewis6991** [comment]
Ah that makes sense as I removed the comment parser.

Why are there even comment injections for vimdoc?

**Comment by clason** [comment]
Not for vimdoc, for vimscript (which is injected into vimdoc now).

File: runtime/queries/vimdoc/injections.scm [hunk 26]
```diff
@@ -1,4 +1,4 @@
 ((codeblock
   (language) @injection.language
   (code) @injection.content)
- (#set! injection.include-children))
```
Comment (clason) [inline runtime/queries/vimdoc/injections.scm:4]: That will break the lua->vim injection in our other docs, which I would hate to lose?

Comment (lewis6991) [reply runtime/queries/vimdoc/injections.scm:4]: Are they really worth it?

The alternative (right now) would be to have no injections.

We can re-enable this when we add lazy injections.

Comment (clason) [reply runtime/queries/vimdoc/injections.scm:4]: Yes, they are really worth it. I'd rather drop the comment parser from nvim-treesitter.

Comment (lewis6991) [reply runtime/queries/vimdoc/injections.scm]: ok

**Comment by clason** [comment]
> Should I wait on this before merging #24508 ?

Depends on how hard it is to rebase this PR (i.e., how much manual work was necessary and has to be repeated). Otherwise there are independent benefits of #24508 (cleaner docs) and no drawbacks, so it would make sense to just go ahead with it already.

(Temporarily disabling the comment->vimscript injections would be the simplest workaround to unblock this.)

File: src/nvim/eval.lua [hunk 27]
```diff
@@ -9335,25 +9320,27 @@ M.funcs = {
...
-      <  - If {address} is omitted the name is "nvim". >
-      	:echo serverstart()
+      <  - If {address} is omitted the name is "nvim". >vim
+      	echo serverstart()
+      < >
       	=> /tmp/nvim.bram/oknANW/nvim.15430.5
+      <
 
-      <Example bash command to list all Nvim servers: >
+      <Example bash command to list all Nvim servers: >vim
```
Comment (clason) [inline src/nvim/eval.lua:9331]:
```suggestion
      <Example bash command to list all Nvim servers: >bash
```

**Comment by lewis6991** [comment]
I say we just go with this. I'm petrified of having to deal with merge conflicts from this.

**Comment by clason** [comment]
Brace yourself then!

**Pull request merged**